### PR TITLE
8262398: Shenandoah: Disable nmethod barrier and stack watermark when running with passive mode

### PR DESCRIPTION
**Pull request by zhengyu123**
nmethod barrier and stack watermark allow GC not to process nmethods at GC pauses, and aim to reduce GC latency, they do not benefit STW GCs, who process nmethods at pauses anyway.

Test:

- [x] hotspot_gc_shenandoah
- [x] tier1 with -XX:+UseShenandoahGC
- [x] tier1 with -XX:+UseShenandoahGC -XX:ShenandoahGCMode=passive
- [x]  tier1 with -XX:+UseShenandoahGC -XX:ShenandoahGCMode=passive -XX:-ShenandoahDegeneratedGC

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262398](https://bugs.openjdk.java.net/browse/JDK-8262398): Shenandoah: Disable nmethod barrier and stack watermark when running with passive mode

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**)
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to 2a45cb02cb28329a128eb8b55ee6f518cd095376

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/2727/head:pull/2727`
`$ git checkout pull/2727`

To update a local copy of the PR:
`$ git checkout pull/2727`
`$ git pull https://git.openjdk.java.net/jdk pull/2727/head`

### CODE REVIEW

**Review by rkennke — COMMENTED** [review]
If it is possible at all, it would be most useful if we could have flags like ShenandoahNMethodBarrier and ShenandoahStackWatermark, and let them drive enabling and disabling of the corresponding barriers/features. It would be even better, if we could configure concurrent modes such that they do STW class-unloading and stack-processing (if possible with reasonable amount of work). That would help ports where those features are not yet present (current ports under development with the problem are PPC and Graal).

**Comment by zhengyu123** [comment]
> If it is possible at all, it would be most useful if we could have flags like ShenandoahNMethodBarrier and ShenandoahStackWatermark, and let them drive enabling and disabling of the corresponding barriers/features. It would be even better, if we could configure concurrent modes such that they do STW class-unloading and stack-processing (if possible with reasonable amount of work). That would help ports where those features are not yet present (current ports under development with the problem are PPC and Graal).

Disabling nmethod barrier and/or stack watermark for concurrent GC means crashes, we no longer have backup (e.g. processing thread roots and code roots at pauses) so I don't think it will help ports.

File: src/hotspot/share/gc/shenandoah/mode/shenandoahPassiveMode.cpp [hunk 1]
```diff
@@ -37,6 +37,14 @@ void ShenandoahPassiveMode::initialize_flags() const {
   // Passive runs with max speed for allocation, because GC is always STW
   SHENANDOAH_ERGO_DISABLE_FLAG(ShenandoahPacing);
 
+  if (FLAG_IS_DEFAULT(ShenandoahNMethodBarrier)) {
+    FLAG_SET_DEFAULT(ShenandoahNMethodBarrier, false);
+  }
+
+  if (FLAG_IS_DEFAULT(ShenandoahStackWatermarkBarrier)) {
+    FLAG_SET_DEFAULT(ShenandoahStackWatermarkBarrier, false);
+  }
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/mode/shenandoahPassiveMode.cpp:46]: This should just go to "// Disable known barriers by default" block below.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 2]
```diff
@@ -402,7 +402,7 @@ jint ShenandoahHeap::initialize() {
   return JNI_OK;
 }
 
-void ShenandoahHeap::initialize_heuristics() {
+void ShenandoahHeap::initialize_gc_mode() {
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:405]: There is `initialize_heuristics`, which means this should be `initialize_mode`?

File: src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp [hunk 3]
```diff
@@ -176,7 +176,7 @@ template <typename IsAlive, typename KeepAlive>
 void ShenandoahRootUpdater::roots_do(uint worker_id, IsAlive* is_alive, KeepAlive* keep_alive) {
   CodeBlobToOopClosure update_blobs(keep_alive, CodeBlobToOopClosure::FixRelocations);
   ShenandoahCodeBlobAndDisarmClosure blobs_and_disarm_Cl(keep_alive);
-  CodeBlobToOopClosure* codes_cl = ClassUnloading ?
+  CodeBlobToOopClosure* codes_cl = ClassUnloading && ShenandoahNMethodBarrier ?
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp:179]: The hunk above has this in parentheses. Pick a style and stick to it, whether parentheses or not.

File: src/hotspot/share/gc/shenandoah/shenandoahCodeRoots.cpp [hunk 4]
```diff
@@ -179,8 +179,11 @@ class ShenandoahDisarmNMethodsTask : public AbstractGangTask {
 };
 
 void ShenandoahCodeRoots::disarm_nmethods() {
-  ShenandoahDisarmNMethodsTask task;
-  ShenandoahHeap::heap()->workers()->run_task(&task);
+  ShenandoahHeap* const heap = ShenandoahHeap::heap();
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahCodeRoots.cpp:182]: This `heap` getter can be moved into the block below, and maybe even inlined?

**Review by shipilev — APPROVED** [review]
Yes, thank you. A few minor nits below.

**Comment by shipilev** [comment]
Forgot one thing. These tests might also need modification to test that we never disable this for true heuristics, etc.

```
./test/hotspot/jtreg/gc/shenandoah/options/TestSelectiveBarrierFlags.java
./test/hotspot/jtreg/gc/shenandoah/options/TestWrongBarrierEnable.java
./test/hotspot/jtreg/gc/shenandoah/options/TestWrongBarrierDisable.java
```

File: test/hotspot/jtreg/gc/shenandoah/options/TestSelectiveBarrierFlags.java [hunk 5]
```diff
@@ -46,6 +46,8 @@ public static void main(String[] args) throws Exception {
                 new String[] { "ShenandoahSATBBarrier", "ShenandoahIUBarrier" },
                 new String[] { "ShenandoahCASBarrier" },
                 new String[] { "ShenandoahCloneBarrier" },
+                new String[] { "ShenandoahNMethodBarrier"},
+                new String[] { "ShenandoahStackWatermarkBarrier"}
```
Comment (shipilev) [inline test/hotspot/jtreg/gc/shenandoah/options/TestSelectiveBarrierFlags.java:50]: Note there is a space between `"` and `}`. Minor nit, fix before push.

**Review by shipilev — APPROVED** [review]

**Review by rkennke — APPROVED** [review]
Looks good! Thanks!

**Pull request closed without merging**